### PR TITLE
Fix fatal error: concurrent map writes for runtime controllers

### LINKED ISSUE #3756: [BUG] Runtime controller exits with fatal error: concurrent map writes
**What is your environment(Kubernetes version, Fluid version, etc.)**

**Describe the bug**
When launching many threads to handle the lifecycles of many Datasets and Runtimes, it is highly possible that the runtime controller exits with the following fatal error:
```
fatal error: concurrent map writes
```

The dumped goroutine is listed:
```
goroutine 716 [running]:
runtime.throw({0x2336787?, 0xc0480e1740?})
	/usr/local/go/src/runtime/panic.go:992 +0x71 fp=0xc00d012580 sp=0xc00d012550 pc=0x43d6b1
runtime.mapdelete_faststr(0x2108180, 0xc00034d980, {0xc0327933b0, 0x2a})
	/usr/local/go/src/runtime/map_faststr.go:386 +0x1aa fp=0xc00d0125f8 sp=0xc00d012580 pc=0x4158ea
github.com/fluid-cloudnative/fluid/pkg/metrics.(*datasetMetrics).Forget(0xc0327a89c0)
	/go/src/github.com/fluid-cloudnative/fluid/pkg/metrics/dataset_metrics.go:71 +0xa5 fp=0xc00d012658 sp=0xc00d0125f8 pc=0x1ceb185
github.com/fluid-cloudnative/fluid/pkg/controllers.(*RuntimeReconciler).ForgetMetrics(0xc0000b41e0, {{0x24eab38, 0xc0574db8c0}, {{0xc0098ea258, 0x11}, {0xc0098ea240, 0x18}}, {0x2326f2d, 0xa}, 0x0, ...})
	/go/src/github.com/fluid-cloudnative/fluid/pkg/controllers/runtime_controller.go:391 +0x114 fp=0xc00d012718 sp=0xc00d012658 pc=0x1f5f194
github.com/fluid-cloudnative/fluid/pkg/controllers.(*RuntimeReconciler).ReconcileRuntimeDeletion(0xc0000b41e0, {0x24ef980, 0xc031bbef00}, {{0x24eab38, 0xc0574db8c0}, {{0xc0098ea258, 0x11}, {0xc0098ea240, 0x18}}, {0x2326f2d, ...}, ...})
	/go/src/github.com/fluid-cloudnative/fluid/pkg/controllers/runtime_controller.go:192 +0x958 fp=0xc00d012d20 sp=0xc00d012718 pc=0x1f5b978
github.com/fluid-cloudnative/fluid/pkg/controllers.(*RuntimeReconciler).ReconcileInternal(0xc0000b41e0, {{0x24eab38, 0xc0574db8c0}, {{0xc0098ea258, 0x11}, {0xc0098ea240, 0x18}}, {0x2326f2d, 0xa}, 0x0, ...})
	/go/src/github.com/fluid-cloudnative/fluid/pkg/controllers/runtime_controller.go:121 +0x19dc fp=0xc00d0132e8 sp=0xc00d012d20 pc=0x1f5addc
github.com/fluid-cloudnative/fluid/pkg/controllers/v1alpha1/jindo.(*RuntimeReconciler).Reconcile(0xc00064d5a0, {0x24eab38, 0xc0574db8c0}, {{{0xc0098ea258, 0x11}, {0xc0098ea240, 0x18}}})
	/go/src/github.com/fluid-cloudnative/fluid/pkg/controllers/v1alpha1/jindo/jindoruntime_controller.go:107 +0xc78 fp=0xc00d0138d0 sp=0xc00d0132e8 pc=0x1f69198
github.com/fluid-cloudnative/fluid/vendor/sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Reconcile(0xc000be6160, {0x24eab38, 0xc0574db8c0}, {{{0xc0098ea258, 0x11}, {0xc0098ea240, 0x18}}})
	/go/src/github.com/fluid-cloudnative/fluid/vendor/sigs.k8s.io/controller-runtime/pkg/internal/controller/controller.go:114 +0x365 fp=0xc00d013a48 sp=0xc00d0138d0 pc=0x1a1b805
github.com/fluid-cloudnative/fluid/vendor/sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).reconcileHandler(0xc000be6160, {0x24eab38, 0xc0574db830}, {0x2179a60, 0xc0164fb620})
	/go/src/github.com/fluid-cloudnative/fluid/vendor/sigs.k8s.io/controller-runtime/pkg/internal/controller/controller.go:311 +0x425 fp=0xc00d013d68 sp=0xc00d013a48 pc=0x1a1e665
github.com/fluid-cloudnative/fluid/vendor/sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).processNextWorkItem(0xc000be6160, {0x24eaa90, 0xc000891d00})
	/go/src/github.com/fluid-cloudnative/fluid/vendor/sigs.k8s.io/controller-runtime/pkg/internal/controller/controller.go:266 +0x398 fp=0xc00d013f28 sp=0xc00d013d68 pc=0x1a1db38
github.com/fluid-cloudnative/fluid/vendor/sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Start.func2.2()
	/go/src/github.com/fluid-cloudnative/fluid/vendor/sigs.k8s.io/controller-runtime/pkg/internal/controller/controller.go:227 +0xdc fp=0xc00d013fe0 sp=0xc00d013f28 pc=0x1a1d23c
runtime.goexit()
	/usr/local/go/src/runtime/asm_amd64.s:1571 +0x1 fp=0xc00d013fe8 sp=0xc00d013fe0 pc=0x4713e1
created by github.com/fluid-cloudnative/fluid/vendor/sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Start.func2
	/go/src/github.com/fluid-cloudnative/fluid/vendor/sigs.k8s.io/controller-runtime/pkg/internal/controller/controller.go:223 +0x4d6
...
goroutine 657 [running]:
runtime.throw({0x2336787?, 0x7f1817f5a818?})
	/usr/local/go/src/runtime/panic.go:992 +0x71 fp=0xc011436538 sp=0xc011436508 pc=0x43d6b1
runtime.mapassign_faststr(0x2108180, 0xc00034d980, {0xc0328ac930, 0x2a})
	/usr/local/go/src/runtime/map_faststr.go:295 +0x406 fp=0xc0114365a8 sp=0xc011436538 pc=0x4156e6
github.com/fluid-cloudnative/fluid/pkg/metrics.GetDatasetMetrics({0xc00b127b30, 0x11}, {0xc00b127b18, 0x18})
	/go/src/github.com/fluid-cloudnative/fluid/pkg/metrics/dataset_metrics.go:54 +0x1d8 fp=0xc011436658 sp=0xc0114365a8 pc=0x1ceaf58
github.com/fluid-cloudnative/fluid/pkg/controllers.(*RuntimeReconciler).ForgetMetrics(0xc0000b41e0, {{0x24eab38, 0xc057550ea0}, {{0xc00b127b30, 0x11}, {0xc00b127b18, 0x18}}, {0x2326f2d, 0xa}, 0x0, ...})
	/go/src/github.com/fluid-cloudnative/fluid/pkg/controllers/runtime_controller.go:391 +0x10a fp=0xc011436718 sp=0xc011436658 pc=0x1f5f18a
github.com/fluid-cloudnative/fluid/pkg/controllers.(*RuntimeReconciler).ReconcileRuntimeDeletion(0xc0000b41e0, {0x24ef980, 0xc03194ec80}, {{0x24eab38, 0xc057550ea0}, {{0xc00b127b30, 0x11}, {0xc00b127b18, 0x18}}, {0x2326f2d, ...}, ...})
	/go/src/github.com/fluid-cloudnative/fluid/pkg/controllers/runtime_controller.go:192 +0x958 fp=0xc011436d20 sp=0xc011436718 pc=0x1f5b978
github.com/fluid-cloudnative/fluid/pkg/controllers.(*RuntimeReconciler).ReconcileInternal(0xc0000b41e0, {{0x24eab38, 0xc057550ea0}, {{0xc00b127b30, 0x11}, {0xc00b127b18, 0x18}}, {0x2326f2d, 0xa}, 0x0, ...})
	/go/src/github.com/fluid-cloudnative/fluid/pkg/controllers/runtime_controller.go:121 +0x19dc fp=0xc0114372e8 sp=0xc011436d20 pc=0x1f5addc
github.com/fluid-cloudnative/fluid/pkg/controllers/v1alpha1/jindo.(*RuntimeReconciler).Reconcile(0xc00064d5a0, {0x24eab38, 0xc057550ea0}, {{{0xc00b127b30, 0x11}, {0xc00b127b18, 0x18}}})
	/go/src/github.com/fluid-cloudnative/fluid/pkg/controllers/v1alpha1/jindo/jindoruntime_controller.go:107 +0xc78 fp=0xc0114378d0 sp=0xc0114372e8 pc=0x1f69198
github.com/fluid-cloudnative/fluid/vendor/sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Reconcile(0xc000be6160, {0x24eab38, 0xc057550ea0}, {{{0xc00b127b30, 0x11}, {0xc00b127b18, 0x18}}})
	/go/src/github.com/fluid-cloudnative/fluid/vendor/sigs.k8s.io/controller-runtime/pkg/internal/controller/controller.go:114 +0x365 fp=0xc011437a48 sp=0xc0114378d0 pc=0x1a1b805
github.com/fluid-cloudnative/fluid/vendor/sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).reconcileHandler(0xc000be6160, {0x24eab38, 0xc057550e10}, {0x2179a60, 0xc0106b5ee0})
	/go/src/github.com/fluid-cloudnative/fluid/vendor/sigs.k8s.io/controller-runtime/pkg/internal/controller/controller.go:311 +0x425 fp=0xc011437d68 sp=0xc011437a48 pc=0x1a1e665
github.com/fluid-cloudnative/fluid/vendor/sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).processNextWorkItem(0xc000be6160, {0x24eaa90, 0xc000891d00})
	/go/src/github.com/fluid-cloudnative/fluid/vendor/sigs.k8s.io/controller-runtime/pkg/internal/controller/controller.go:266 +0x398 fp=0xc011437f28 sp=0xc011437d68 pc=0x1a1db38
github.com/fluid-cloudnative/fluid/vendor/sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Start.func2.2()
	/go/src/github.com/fluid-cloudnative/fluid/vendor/sigs.k8s.io/controller-runtime/pkg/internal/controller/controller.go:227 +0xdc fp=0xc011437fe0 sp=0xc011437f28 pc=0x1a1d23c
runtime.goexit()
	/usr/local/go/src/runtime/asm_amd64.s:1571 +0x1 fp=0xc011437fe8 sp=0xc011437fe0 pc=0x4713e1
created by github.com/fluid-cloudnative/fluid/vendor/sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Start.func2
	/go/src/github.com/fluid-cloudnative/fluid/vendor/sigs.k8s.io/controller-runtime/pkg/internal/controller/controller.go:223 +0x4d6
```

**What you expect to happen:**
Runtime controller should not fail with fatal error.

**How to reproduce it**
Set `--runtime-workers` to a large number (e.g. 40), and create Datasets and Runtimes in a fast pace may trigger this problem.

**Additional Information**

### PR DESCRIPTION
**Pull request by TrafalgarZZZ**
### Ⅰ. Describe what this PR does
#3756 is caused by concurrent writes on Fluid's metrics maps. Specifically, when there are multiple goroutines reconciling multiple Runtimes, it is possible that a deletion and a inserting happen at the same time on the same metrics map, which makes `fatal error: concurrent map writes`.

This PR fixes by adding Mutex to protect race condition on the maps.

### Ⅱ. Does this pull request fix one issue?

fixes #3756 

### Ⅴ. Special notes for reviews

### CODE REVIEW

File: pkg/metrics/dataset_metrics.go [hunk 1]
```diff
@@ -74,4 +83,5 @@ func (m *datasetMetrics) Forget() {
 func init() {
 	metrics.Registry.MustRegister(datasetUFSFileNum, datasetUFSTotalSize)
 	datasetMetricsMap = map[string]*datasetMetrics{}
+	datasetMetricsMutex = &sync.Mutex{}
```
Comment (cheyang) [inline pkg/metrics/dataset_metrics.go:86]: Why not using `&sync.RWMutex{}` to separate read and write?

Comment (TrafalgarZZZ) [reply pkg/metrics/dataset_metrics.go:86]: The two related operations are insertion and deletion. There is no read operation on metrics map.

Comment (TrafalgarZZZ) [reply pkg/metrics/dataset_metrics.go:86]: I see what you mean. RWMutex lgtm, will change it.

**Review by cheyang — APPROVED** [review]
/lgtm
/approve

**Pull request merged**